Add JPEG encoder support via new camera_encoder component

DT-art1 opened the pull request:
# What does this implement/fix?

This introduces a new camera_encoder component. A default software JPEG encoder implementation is included, allowing the camera component to compress and send images in the expected JPEG format to API clients.

- This is the second step in a series of changes aimed at modernizing and generalizing ESPHome's camera support.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes #7639

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5314

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
camera_encoder:
  id: jpeg_encoder
  quality: 80
  buffer_size: 4096
  buffer_expand_size: 1024
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

DT-art1 committed:
Add new camera_encoder component with default JPEG encoder

This introduces a new camera_encoder component. A default software JPEG
encoder implementation is included, allowing the camera component to
compress and send images in the expected JPEG format to API clients.

DT-art1 committed:
Updated CODEOWNERS

bdraco commented:
I'll try to find some time to review next week after the release work calms down

DT-art1 committed:
Cleanup configuration.

bdraco commented on esphome/components/camera_encoder/__init__.py:
```diff
@@ -0,0 +1,67 @@
+
+async def to_code(config):
+    cg.add_build_flag("-DUSE_CAMERA_SW_JPEG_ENCODER")
+    cg.add_library("dt-art1/jpegenc-pio", "1.0.0")
```
https://registry.platformio.org/libraries/dt-art1/jpegenc-pio

I see you are using a fork of https://github.com/bitbank2/JPEGENC. Was there an attempt made to get the changes merged upstream and rejected?

DT-art1 commented:
The first PR https://github.com/bitbank2/JPEGENC/pull/8 was successfully merged upstream. However, the second PR https://github.com/bitbank2/JPEGENC/pull/9 ,  which fixes a color ordering issue for RGB and adds support for additional color formats, was not merged. I was unable to get any response from the original author despite several attempts, so I eventually closed it.

To proceed with the development of the modular camera framework, I decided to fork the library. JPEGENC is a solid candidate for a default software encoder for various software cameras, and forking it enabled continued development without blocking.

I would be happy it the ESPHome team wants to publish the fork to the PlatformIO registry to make integration cleaner and more official.

Support for the ESP32-P4's hardware JPEG encoder can be added later, and that's the reasoning behind the DEFAULT_ENCODER configuration parameter - to allow for selecting platform-specific encoders while keeping the software fallback available.

DT-art1 committed:
Merge branch 'dev' into camera_encoder

DT-art1 committed:
Added ESP32Camera JPEG encoder as default encoder.

- Replaced RGB with BGR, as both esp32_camera and P4 use BGR byte order
internally.
- Cleaned up doxygen comments.
- Simplified and cleaned up the encoder interface.

DT-art1 commented:
Kill your darlings...

Considering the low likelihood that the changes in bitbank2 will ever be accepted upstream, I've removed it and instead integrated the JPEG encoder from esp32_camera. This encoder requires no new additional library and should be straightforward to integrate.

I haven't added incremental encoding support yet, as that can be introduced later. My hope is that replacing the bitbank2 encoder with the one from esp32_camera helps move this PR forward.

RubenKelevra commented:
I mean this has also advantages:

I took a look at the encoder today and I think there's a lot of potential to speed it up, with highly tailored versions for each platform. So the precompiler could just select more optimized sections based on the platform we're compiling for.

Especially the S3 could quite some benefits from that.

RubenKelevra commented:
> I haven't done a proper performance comparison, but based on side-by-side testing, I don't feel the difference is that significant.

No, I agree, there won't be much of a difference between two JPEG implementations.

The things which can however speed the encoder up on the S3 platform is that it supports vector operations, got DSP operations and a floating point co-processor.

The code makes no use of that and the compiler obviously has no idea how to use that either, except maybe floating point.

But since not all platforms got floating point operations switching in general to them would slow the encoder down for those platforms, as it would need to be emulated in a software library, which is a jump, instead of staying in a tight loop in the encoder library.

The hardware support matrix is kinda mixed here:

The original ESP32 does support Floating Point operations in hardware, but the S2 does not, while the S2 got more powerful DSP operations, which the ESP32 can only do shorter chains of etc.

And the new line of RISC-V has basically nothing which helps us, nothing even an FPU.

Im thus building now an optimized version which is build out of blocks by the precompiler to build exactly the right optimization for the hardware in question in, and default to the pure original Cpp version if the platform is not known.

DT-art1 commented:
Makes sense - glad we're aligned on this! Looking forward to what you build with the precompiler blocks. :+1:

RubenKelevra commented:
Yeah I'm half through my todo, should be done by tomorrow.

RubenKelevra commented:
Yeah, this will take a bit longer I guess. Progress has been slow today 🥲

I'll keep you posted!

bdraco commented:
I'll take a look as soon as I have some free cycles.. Likely after beta and release

DT-art1 commented:
Sounds good, thanks for letting me know!

RubenKelevra commented:
@DT-art1 similar for my progress, the updater issues (https://github.com/esphome/esphome/pull/10182) and weird network behavior of Linux pushed in front of my work on the optimizations. Will continue on once I'm done with the Kernel fix.

DT-art1 commented:
@RubenKelevra I have a related question: Would it be feasible to add incremental encoding support to esp32_camera?
One big benefit would be avoiding long blocking periods during encoding - for example, keeping each processing loop call under 30ms to stay within ESPHome's time constraints.

The bitbank2 encoder supports chunk/incremental encoding, which is helpful here.
Do you think something like that could be integrated into the current design, or would it require a major refactor ?

RubenKelevra commented:
@DT-art1 I mean there's technically no reason why it shouldn't work. The constraint makes especially sense on single core devices. Dual core devices (ESP32 / ESP32-S3) does network stuff on the core 0 while ESPHome stuff mainly runs on core 1.

So the hit on communication latency spikes is less severe from my understanding.

The performance expectations *should* be similar to the Espressif closed source JPEG encoder (`esp_new_jpeg` great naming btw 🥲) which we could leverage on some situations, because it only accepts RGB888 (and some more formats).

See performance numbers [here](https://components.espressif.com/components/espressif/esp_new_jpeg). 

_Not sure if I can reach the same level of optimizations here, thought. So don't keep your hopes too much up here._ 

If the device gives us RGB565 it would be quite slow if we first need to expand the bit depth, before we can throw it in. Apart from that I dislike the closed source approach here. I'm leveraging the ESP-DSP library, which is also closed source, but only optional. So it's user configurable if closed source is used or not. So all code paths are also available as open source.

Since we do a lot of looping we can surely jump out of them for short whiles, but this will have some impact on the performance I think.

The reason here is, that doing other stuff could lead to another flash read for our functions or that other stuff is reading from the PSRAM and thus our already cached blocks drop from the cache. 

So it only makes sense if we would really exceed the time limit. So on small images it would make more sense to keep the tight loop.


I give this some thought.

DT-art1 commented:
Yes - right now, there is a single buffer allocated and assigned to the encoder as its output buffer. The encoder fills the buffer, and when done, both the buffer and its length are passed to the API to serve the clients.

I don't see any problem with adding additional buffers to the encoder in the future so it could switch slots and start encoding the next image immediately. However, there may be other ongoing tasks in parallel - not only the camera pipeline (capturing, overlaying text, image rescaling, colorization) but also other subsystems. For example, I'm currently running this alongside a voice assistant, using an MLX90640 as a presence sensor, and it works well.

For this reason, I think double buffering could be added later, once the essential parts are in ESPHome. My preference is to first get the camera pipeline merged, as it provides broad value for others. If this encoder makes it into ESPHome, it opens the door for different applications to send images into Home Assistant and other clients.

Once the camera pipeline is in place, the camera can be integrated into anything as it becomes a platform component - making it simple to visualize all sorts of data. While thermal cameras are currently my main focus, there are many other sensors that could benefit from this, such as radar sensors - and perhaps even LVGL in the future.

RubenKelevra commented:
Well, your explanation and my explanation are not the same:

In my case the encoder would be tasked with creating the memory slots and you would get the pointer+image length at the end of the encoding, from the encoder, and just return the slot if your application is done reading.

This makes sense because the optimum memory alignment depends on ESP-IDF versions and settings, which means the encoder needs to make sure the memory is properly aligned for optimum performance.

Same goes for the input image actually, we need this to be properly aligned, too, for optimal performance.

So the "least work for you" to do proper alignment and handle double buffering would be, to specify what format and resolution the source image got and init the encoder with that. This will create the two input and output buffers, so input 1 -> output 1 and input 2 -> output 2.

Once you've written input 1 you start encoding, which is basically returning the handle, so you're not allowed to do further writes. If you wrote via DMA from a camera to the PSRAM you can specify that on the return and the necessary cache drop will be handled as well.

At the end of the encoding you can grab the output handle of output 1 and eventually return it.

This means you just have to "open", "write", "close" the inputs, do an encode and once it's done you "open", "read", "close".

The encoder would make sure, that you cannot start an encoding while the output is still "open".

This means you can encode one image in a dedicated job while another job is already fetching the next image from the camera into the second slot, while the second output slot is still read from the application - if this makes sense.

DT-art1 commented:
Yes, PR #7639 already shows that letting the encoder allocate the memory is straightforward (esp_p4_encoder_buffer). That's exactly why EncoderBuffer is defined as an interface.

I'm not yet sure whether it's better in the long run to exten RAMAllocator with alignment support or to use a specialized EncoderBuffer implementation - but both options are possible. From a functional perspective, nothing changes, though performance could be improved later.

Since smaller, incremental changes are preferred, I'd suggest leaving such optimizations for a later PR. At the moment, I don't see anything that would block this from being added in the future.

Thanks for all your input!

RubenKelevra commented:
> I'm not yet sure whether it's better in the long run to exten RAMAllocator with alignment support

I think in the process of including this into esp32-camera driver I will likely make a small lib out of the alignment and cache drop functionality, which is esp-idf version independent then. Meaning it choses the right functions to fetch the settings for the right hardware/esp-idf version automatically.

Because I need it both in the cam_hal.c as well as the jpeg_encoder.

So maybe you can just include this function as well then, if you need something like that?

DT-art1 commented:
Yes, sounds good. I could then replace the default EncoderBuffer with a specialized one using your new library - or perhaps use it instead of the current RAMAllocator.
We'll see when we get there.

DT-art1 committed:
Added generic Buffer.

Introduces a generic buffer abstraction to decouple image data handling
from CameraRequester, avoiding unnecessary dependencies.

DT-art1 commented:
The main #7639 has progressed, and the encoder-related parts should no longer change. I just updated this PR with the latest changes from #7639. It turned out that introducing an additional **Buffer** simplifies the camera pipeline, since it removes the dependency on **CameraRequester**.

There may be some performance improvements later in the esp32_camera library (external to ESPHome). If those land, minor adjustments might be needed, but I don't expect them to affect the encoder interface.

bdraco commented on esphome/components/camera/buffer.h:
```diff
@@ -0,0 +1,20 @@
+#include <cinttypes>
+#include <cstddef>
+
+namespace esphome {
+namespace camera {
```
```suggestion
namespace esphome::camera {
```

bdraco commented on esphome/components/camera/buffer.h:
```diff
@@ -0,0 +1,20 @@
+};
+
+}  // namespace camera
+}  // namespace esphome
```
```suggestion
}  // namespace esphome::camera
```

bdraco committed:
Merge branch 'dev' into camera_encoder

bdraco committed:
name space cleanups as we have c++17 now (c++20 actually)

bdraco committed:
name space cleanups as we have c++17 now (c++20 actually)

bdraco committed:
name space cleanups as we have c++17 now (c++20 actually)

bdraco committed:
name space cleanups as we have c++17 now (c++20 actually)

bdraco committed:
name space cleanups as we have c++17 now (c++20 actually)

bdraco committed:
name space cleanups as we have c++17 now (c++20 actually)

bdraco committed:
name space cleanups as we have c++17 now (c++20 actually)

bdraco committed:
reorder members to reduce padding

bdraco commented on esphome/components/camera_encoder/esp32_camera_jpeg_encoder.cpp:
```diff
@@ -0,0 +1,82 @@
+void ESP32CameraJPEGEncoder::dump_config() {
+  ESP_LOGCONFIG(TAG,
+                "ESP32 Camera JPEG Encoder:\n"
+                "  Size: %d\n"
```
```suggestion
                "  Size: %zu\n"
```

bdraco committed:
Update esphome/components/camera_encoder/esp32_camera_jpeg_encoder.cpp

bdraco commented on esphome/components/camera_encoder/__init__.py:
```diff
@@ -0,0 +1,56 @@
+        cv.GenerateID(): cv.declare_id(ESP32CameraJPEGEncoder),
+        cv.Optional(CONF_QUALITY, default=80): cv.int_range(1, 100),
+        cv.Optional(CONF_BUFFER_SIZE, default=4096): cv.int_range(1024),
+        cv.Optional(CONF_BUFFER_EXPAND_SIZE, default=1024): cv.int_range(0),
```
Please set a min/max here

DT-art1 replied:
Thanks for pointing that out! I've added explicit min/max limits for the JPEG buffer. The upper bound is set to 2 MB, which should be sufficient for ESP32-S3 and ESP32-P4.

DT-art1 committed:
Rename ImageFormat to PixelFormat

The previous name was misleading: GRAYSCALE, RGB565, and BGR888 describe
the pixel representation rather than the image as a whole. The name
ImageFormat may later be reused to distinguish between encoded (e.g.
JPEG) and unencoded formats in the camera pipeline.

DT-art1 committed:
Add min/max limits for JPEG buffer allocation

Introduced explicit lower and upper bounds for JPEG buffer allocation
and resizing. The upper limit (2 MB) should be sufficient for most use
cases on ESP32-S3 and ESP32-P4, while still preventing excessive memory
usage.
